### PR TITLE
ADR-1194 Add data structures, config and connector for returns submission

### PR DESCRIPTION
**Pull request by davidfes**

### CODE REVIEW

File: app/uk/gov/hmrc/alcoholdutyreturns/models/returns/AdrReturnsSubmission.scala [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+                        spoiltProduct: BigDecimal,
+                        drawback: BigDecimal,
+                        repackagedDraught: BigDecimal,
+                        totalDutyDue: BigDecimal
+                      )
+
+case object AdrTotals {
+  implicit val adrTotalsReads: Reads[AdrTotals] = Json.reads[AdrTotals]
+}
+
```
Comment (davidfes) [inline app/uk/gov/hmrc/alcoholdutyreturns/models/returns/AdrReturnsSubmission.scala:244]: Think the singular AdrReturnSubmission sounds better here (+scala file rename)

Comment (davidfes) [reply app/uk/gov/hmrc/alcoholdutyreturns/models/returns/AdrReturnsSubmission.scala]: Done

File: test/uk/gov/hmrc/alcoholdutyreturns/repositories/CacheRepositorySpec.scala [hunk 2]
```diff
@@ -1,67 +0,0 @@
-/*
```
Comment (davidfes) [inline test/uk/gov/hmrc/alcoholdutyreturns/repositories/CacheRepositorySpec.scala:1]: Test was flaky, plus it's not a unit test if it's accessing Mongo. We have a repo test in 'it' already.

File: app/uk/gov/hmrc/alcoholdutyreturns/controllers/ReturnsController.scala [hunk 3]
```diff
@@ -51,6 +53,22 @@ class ReturnsController @Inject() (
         )
     }
 
+  def submitReturn(appaId: String, periodKey: String): Action[JsValue] =
+    authorise(parse.json).async { implicit request =>
+      withJsonBody[AdrReturnSubmission] { returnSubmission =>
+        returnsService
+          .submitReturn(returnSubmission, ReturnId(appaId, periodKey))
```
Comment (davidfes) [inline app/uk/gov/hmrc/alcoholdutyreturns/controllers/ReturnsController.scala:60]: Should validate that if a section is not filled in, the items are not present (Seq/Set being empty)

File: app/uk/gov/hmrc/alcoholdutyreturns/connector/CalculatorConnector.scala [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+  )(implicit hc: HeaderCarrier, writes: Writes[I], reads: Reads[O]): EitherT[Future, ErrorResponse, O] =
+    EitherT(
+      httpClient
+        .POST[I, Either[UpstreamErrorResponse, HttpResponse]](
+          url = url,
+          body = requestBody
+        )
+        .map {
+          case Right(response)                                                =>
+            Try(response.json.as[O]).toOption
```
Comment (yasaracar) [inline app/uk/gov/hmrc/alcoholdutyreturns/connector/CalculatorConnector.scala:51]: As mentioned in team channel, this loses the validation error context. You can use something like:

      Try {
              response.json
                .as[FinancialTransactionDocument]
            } match {
              case Success(doc)       =>
                logger.info(s"Retrieved financial transaction document for appaId $alcoholDutyReference")
                Some(doc)
              case Failure(exception) =>
                logger.warn(
                  s"Parsing failed for financial transaction document with appaId $alcoholDutyReference",
                  exception
                )
                None
            }

File: app/uk/gov/hmrc/alcoholdutyreturns/connector/ReturnsConnector.scala [hunk 5]
```diff
@@ -62,4 +62,31 @@ class ReturnsConnector @Inject() (
...
+  ): EitherT[Future, ErrorResponse, ReturnCreatedDetails] = EitherT(
+    httpClient
+      .POST[ReturnCreate, Either[UpstreamErrorResponse, HttpResponse]](
+        url = config.submitReturnUrl,
+        body = returnToSubmit,
+        headers = headers.submitReturnHeaders(appaId)
+      )
+      .map {
+        case Right(response)                                                =>
+          Try(response.json.as[ReturnCreatedSuccess]).toOption
```
Comment (yasaracar) [inline app/uk/gov/hmrc/alcoholdutyreturns/connector/ReturnsConnector.scala:77]: As mentioned in team channel, this loses the validation error context. You can use something like:

      Try {
              response.json
                .as[FinancialTransactionDocument]
            } match {
              case Success(doc)       =>
                logger.info(s"Retrieved financial transaction document for appaId $alcoholDutyReference")
                Some(doc)
              case Failure(exception) =>
                logger.warn(
                  s"Parsing failed for financial transaction document with appaId $alcoholDutyReference",
                  exception
                )
                None
            }

**Review by yasaracar — CHANGES_REQUESTED** [review]

File: app/uk/gov/hmrc/alcoholdutyreturns/controllers/ReturnsController.scala (hunk 3 above)
Comment (davidfes) [reply app/uk/gov/hmrc/alcoholdutyreturns/controllers/ReturnsController.scala:60]: Validation will happen in a subsequent jira.

**Review by yasaracar — APPROVED** [review]

**Pull request merged**